src/node_http2.cc

```diff
@@ -1364,16 +1364,36 @@ void Http2Session::MaybeScheduleWrite() {
 // storage for data and metadata that was associated with these writes.
 void Http2Session::ClearOutgoing(int status) {
   CHECK_NE(flags_ & SESSION_STATE_SENDING, 0);
-  flags_ &= ~SESSION_STATE_SENDING;
 
-  for (const nghttp2_stream_write& wr : outgoing_buffers_) {
-    WriteWrap* wrap = wr.req_wrap;
-    if (wrap != nullptr)
-      wrap->Done(status);
+  if (outgoing_buffers_.size() > 0) {
+    for (const nghttp2_stream_write& wr : outgoing_buffers_) {
+      WriteWrap* wrap = wr.req_wrap;
+      if (wrap != nullptr)
+        wrap->Done(status);
+    }
+
+    outgoing_buffers_.clear();
+    outgoing_storage_.clear();
```

Comment:
Yeah, it's fine either way. I've moved `flags_ &= ~SESSION_STATE_SENDING;` after this block though so now it's impossible to start another `SendPendingWrite` until after this whole block runs.

I think this way is slightly more clear in terms of all the stuff going on?

Reply:
To elaborate on this, the way I did it before was necessary until I changed the behaviour of `.close()` to wait for the write queue to flush. It used to break our tests. Now our tests are passing and the code is IMO clearer, plus we do a bit less work since we don't need to call `swap`. (And there's less indirection with Done triggering another `SendPendingData`.)

Reply:
I think this is fine, but I also don’t thing there’s anything wrong with clearing `outgoing_storage_` early in this function – the worst thing that could happen is that it frees up memory to be re-used for allocations from the other work done here.